Separate playbooks and merge external roles.

UmanShahzad opened the pull request:
Needed to run a custom playbook for all Open edX appservers.

Since the custom playbook needed Consul & Prometheus' Node Exporter, I had to somehow install these dependencies. I tried doing it through Ocim but it was too difficult because too much test code had to be changed. Since we've had discussions about consolidating our roles to streamline our Ansible-related dev work, I thought this'd be a good opportunity to take a first step and save myself time on this ticket and for others in the future.

I have some more roles to migrate before this can work though...

UmanShahzad commented:
@smarnach BTW I did correctly merge the git trees this time :p Using `git blame` does show the right history.

UmanShahzad commented on .gitignore:
```diff
@@ -1,3 +1,7 @@
+playbooks/roles/*
+!playbooks/roles/consul
+!playbooks/roles/node-exporter
+!playbooks/roles/ernestas-poskus.prometheus
```
Yep, eventually we'll move everything to submodules, which'll allow all roles to be in this repo at all times. It's the cleanest way to avoid a hybrid situation.

UmanShahzad commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "playbooks/roles/ernestas-poskus.prometheus"]
+	path = playbooks/roles/ernestas-poskus.prometheus
+	url = https://github.com/ernestas-poskus/ansible-prometheus
```
I haven't committed it yet, but it's simply a use of GitPython's `Repo.submodule_update()`. It has no effect if there's no submodules, so we can use it for any case without checking for existing submodules (which if we had to, would be `if repo.submodules`).

smarnach replied:
I see, thanks!

UmanShahzad commented on ansible.cfg:
```diff
@@ -1,4 +1,5 @@
 inventory = hosts
-timeout=30
+timeout = 30
+remote_user = ubuntu
```
:+1:

UmanShahzad commented on playbooks/appserver.yml:
```diff
@@ -0,0 +1,20 @@
+# Playbook to set up common services on any type of appserver.
+
+# Temporary: until we migrate all roles in all playbooks from `deploy-all.yml`, we install
+# Python here separately. Otherwise Ocim will complain about missing dependencies.
```
It's actually coming before, which is the cause of the issue. I wanted to do it before to keep it consistent with the approach we took with common-server having consul and the node exporter in them. Do you think we should do it after or?

smarnach replied:
We need to run it after edx_sandbox – otherwise neither Python 2 nor nginx will be available.  You found a fix for the lack of Python 2, but there is no easy fix for the missing nginx.

UmanShahzad replied:
Ah, right. Damn :) I'll make that change soon. As a side note for the future -- once we consolidate roles, should we try moving that `nginx.yml` in common-server into its own role?

smarnach commented on playbooks/dalite.yml:
```diff
@@ -0,0 +1,24 @@
+      # Backup of logs is initiated by logrotate, which incidentally
+      # also creates data to be backed up.
+      TARSNAP_CRONTAB_STATE: "absent"
+    - forward-server-mail
```
I looked into porting to the latest version of Ansible.  Ansible 2.5 was released a month ago, and now has official Python 3 support.  If we migrate, we can use the `import_role` and `import_playbook` modules, which will make some of the semantics of including roles less weird (like the way in which tags and variables are inherited).  Once you have finished and merged this PR, I will give porting to Ansible 2.5.1 a shot – we will also be able to remove the Python 2 installation task.  And I also can't wait to merge all the other roles into this repo as well – I just wish I had done it before last week.

UmanShahzad replied:
OK that sounds like a plan. Note that I may just merge all other roles right here before I want to merge this PR. It's quite mechanical and boring but I can probably write a script and just get it done quickly.

smarnach replied:
I can also merge the other roles, since I have some time available, and I would like to get this PR merged as soon as possible, so I can base any future work on the new version.  Here's what I propose:

* Third-party roles to be inlcuded as git submodules:
  * geerlingguy.security
  * kamaln7.swapfile
  * pmbauer.tarsnap  (Side note: I want to get rid of that role – Tarsnap can now be installed via apt)
  * greendayonfire.mongodb
  * geerlingguy.mysql
  * geerlingguy.postgresql
* OpenCraft roles to be included with git subtree:
  * load-balancer – we have CircleCI set up for this role, so we should keep the repo, but reduce the overhead by using subtree instead of submodule
* OpenCraft roles to be included directly:
  * backup-to-tarsnap
  * common-server
  * configure-apt
  * sanity-checker
  * forward-server-mail
  * backup-swift-container
  * dalite
  * rabbitmq
  * elasticsearch
  * postfix
  * mattermost
  * crafty
  * link-shortener
  * accounting
  * mysql
  * postgres
  * mongodb

I could prepare a PR against your branch if you want me to.

UmanShahzad replied:
OK, I'll just go with submodules. I've also just gone ahead and done it now.

UmanShahzad commented on deploy-all.yml:
```diff
@@ -5,115 +5,16 @@
-    - forward-server-mail
-    - accounting
+- include: playbooks/accounting.yml
+- include: playbooks/appserver.yml
```
Yes for consistency, but I think you're right, this makes sense to just get run directly from Ocim. I don't see it being run for other purposes (except us manually doing it for AWS, but it's good in that case to be explicit).

UmanShahzad commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
 ansible==2.3.2
 netaddr==0.7.19
+mock==2.0.0
+ddt==1.1.2
```
Woops, thanks!

smarnach replied:
Just wondering how this got here, if you didn't do it intentionally.  You do know about the `subtree` merge strategy in Git, right?  (I don't mean the `git subtree` command, which is unrelated.  I mean `git merge -s subtree`.)

UmanShahzad replied:
I had no clue about the subtree merge strategy, no. I was doing it with the default strategy. The requirements.txt file I believe had a conflict with the one at the root and I just manually fixed it, and left it in there. Did this cause something weird we'll have to undo?

smarnach replied:
No, it probably didn't cause any issues except for this one.  It could just have made your live a lot easier, since it's designed to merge another repo as a subtree into the current repo.  The `git subtree` command is basically just a more user-friendly fronted, and we could actually have used it as well:

    git subtree add --prefix=playbooks/roles/whatever https://github.com/open-craft/whatever master

While `git subtree` also allows to later push back to the original repo, it actually copies the full history into the current repo, and it's completely transparent to users, so this would probably have been the easiest option.  But that's for next time. :)

UmanShahzad commented:
@smarnach Yeah we should probably not go further here except for what's already done, and to make it work. I'll include your changeset to move deploy-all.yml to playbooks/, but I don't have a preference if we move the playbooks inside of `playbooks/` out to the root directory -- I only thought it'd get a bit messy.

UmanShahzad commented:
@smarnach Sorry this PR is such a big mess. I've addressed the review and I believe it's ready for another round. I may have to make a small change in documentation for ansible-secrets before merging that part but this can go first.
